tests/test_birkhoff.py

```diff
@@ -0,0 +1,35 @@
+import geoopt
+import torch
+import numpy as np
+import pytest
+
+
+"""
+    This file puts the Birkhoff Polytope, the manifold of doubly stochastic matrices, to test.
+"""
+
+@pytest.mark.parametrize("params", [dict(lr=1e-2), ])
+def test_adam_birkhoff(params):
+    birkhoff = geoopt.manifolds.BirkhoffPolytope()
+    torch.manual_seed(42)
+    with torch.no_grad():
+        X = geoopt.ManifoldParameter(torch.rand(10, 5, 5), manifold=birkhoff).proj_()
+    Xstar = torch.rand(10, 5, 5)
+    Xstar.set_(birkhoff.projx(Xstar))
+
+    optim = geoopt.optim.RiemannianAdam([X], stabilize=4500, **params)
+    for _ in range(10000):
+        optim.zero_grad()
+        loss = torch.norm(X - Xstar)
```

Comment:
In Stiefel manifold, we also add an unreasonably high penalty for violating the manifold constraint to verify numerical stability of optimization. In your case, you better add a similar penalty.